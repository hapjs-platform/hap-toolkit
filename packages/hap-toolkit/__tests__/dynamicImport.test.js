/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

const fs = require('fs')
const path = require('path')
const del = require('del')
const { copyApp, wipeDynamic } = require('hap-dev-utils')
const { compile } = require('../lib')

function testContentMatch(stats, projectRoot) {
  projectRoot = path.resolve(projectRoot)
  const projectRootReg = new RegExp(projectRoot, 'g')
  const json = stats.toJson({ source: true })
  json.modules
    .filter((mod) => mod.source)
    .forEach((module) => {
      expect(wipeDynamic(module.source, [[projectRootReg, '<project-root>']])).toMatchSnapshot()
    })
}

describe('use dynamic import in project', () => {
  const platform = 'native'
  const mode = 'dev'
  let projectRoot
  let manifestPath

  beforeAll(async () => {
    const testAppDir = path.resolve(__dirname, '../fixtures/dynamicImport-app')
    projectRoot = await copyApp(testAppDir)
    manifestPath = path.join(projectRoot, 'src/manifest.json')
  })

  // build 模式添加 --split-chunks-mode=smart 参数时，需要把公共js或组件抽取出来
  it('build results include split chunks for [SMART]', async () => {
    const { stats } = await compile(platform, mode, false, {
      cwd: projectRoot,
      splitChunksMode: 'SMART'
    })
    expect(stats.hasErrors()).toBe(false)
    testContentMatch(stats, projectRoot)
  }, 50000)

  // 动态引用js + 分包
  it('dynamic import with subpackages', async () => {
    const manifest = require(manifestPath)
    manifest.subpackages = [
      {
        name: 'about',
        resource: 'About'
      }
    ]
    fs.writeFileSync(manifestPath, JSON.stringify(manifest))

    const { stats } = await compile(platform, mode, false, {
      cwd: projectRoot,
      splitChunksMode: 'SMART'
    })
    expect(stats.hasErrors()).toBe(false)
  }, 50000)

  afterAll(async () => {
    await del([projectRoot])
  })
})
