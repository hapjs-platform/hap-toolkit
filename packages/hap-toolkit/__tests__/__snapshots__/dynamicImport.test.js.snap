// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`use dynamic import in project build results include split chunks for [SMART] 1`] = `
"
var $app_style$ = {}
var $app_script$ = require(\\"!../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../../node_modules/@hap-toolkit/packager/lib/loaders/module-loader.js!../../../../../node_modules/@hap-toolkit/packager/lib/loaders/manifest-loader.js?path=<project-root>/src!../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&comments=false&configFile=<project-root>/babel.config.js!../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./app.ux?uxType=app\\")

$app_define$('@app-application/app', [], function ($app_require$, $app_exports$, $app_module$) {
  
  $app_script$($app_module$, $app_exports$, $app_require$)
  if ($app_exports$.__esModule && $app_exports$.default) {
    $app_module$.exports = $app_exports$.default
  }
  $app_module$.exports.manifest = require(\\"./manifest.json\\")
  $app_module$.exports.style = { list: [ $app_style$ ] }
  
})
$app_bootstrap$('@app-application/app', { packagerVersion: QUICKAPP_TOOLKIT_VERSION })
"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 2`] = `
"
var $app_script$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../../../node_modules/@hap-toolkit/packager/lib/loaders/module-loader.js!../../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&plugins[]=<CWD>/packages/hap-dsl-xvm/lib/loaders/babel-plugin-jsx.js&comments=false&configFile=<project-root>/babel.config.js!../../../../../hap-dsl-xvm/lib/loaders/access-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./index.ux?uxType=page\\")
$app_define$('@app-component/index', [], function($app_require$, $app_exports$, $app_module$) {
     $app_script$($app_module$, $app_exports$, $app_require$)
        if ($app_exports$.__esModule && $app_exports$.default) {
          $app_module$.exports = $app_exports$.default
        }
    $app_module$.exports.template = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/template-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=template!./index.ux?uxType=page&\\")
    $app_module$.exports.style = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/style-loader.js?index=0&type=style!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=style!./index.ux?uxType=page\\")
});
$app_bootstrap$('@app-component/index',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION });"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 3`] = `
"
var $app_script$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../../../node_modules/@hap-toolkit/packager/lib/loaders/module-loader.js!../../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&plugins[]=<CWD>/packages/hap-dsl-xvm/lib/loaders/babel-plugin-jsx.js&comments=false&configFile=<project-root>/babel.config.js!../../../../../hap-dsl-xvm/lib/loaders/access-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./index.ux?uxType=page\\")
$app_define$('@app-component/index', [], function($app_require$, $app_exports$, $app_module$) {
     $app_script$($app_module$, $app_exports$, $app_require$)
        if ($app_exports$.__esModule && $app_exports$.default) {
          $app_module$.exports = $app_exports$.default
        }
    $app_module$.exports.template = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/template-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=template!./index.ux?uxType=page&\\")
    $app_module$.exports.style = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/style-loader.js?index=0&type=style!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=style!./index.ux?uxType=page\\")
});
$app_bootstrap$('@app-component/index',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION });"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 4`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
require(\\"./global\\");
var _default = {};
exports.default = _default;}"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 5`] = `
"{
  \\"package\\": \\"com.application.demo\\",
  \\"name\\": \\"dynamicImport-app\\",
  \\"versionName\\": \\"1.0.0\\",
  \\"versionCode\\": \\"1\\",
  \\"minPlatformVersion\\": 1070,
  \\"icon\\": \\"/Common/logo.png\\",
  \\"deviceTypeList\\": [\\"phone\\"],
  \\"features\\": [],
  \\"permissions\\": [
    {
      \\"origin\\": \\"*\\"
    }
  ],
  \\"config\\": {
    \\"logLevel\\": \\"log\\"
  },
  \\"router\\": {
    \\"entry\\": \\"Demo\\",
    \\"pages\\": {
      \\"Demo\\": {
        \\"component\\": \\"index\\"
      },
      \\"About\\": {
        \\"component\\": \\"index\\"
      }
    }
  }
}
"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 6`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {
  private: {
    time: new Date().toLocaleString()
  },
  dynamicImportJs() {
    if (this._loaded) {
      this.time = this.timerGenerator();
      return;
    }
    import('./timer').then(res => {
      this.timerGenerator = res.default;
      this._loaded = true;
      this.time = this.timerGenerator();
    });
  }
};
exports.default = _default;
const moduleOwn = exports.default || module.exports;
const accessors = ['public', 'protected', 'private'];
if (moduleOwn.data && accessors.some(function (acc) {
  return moduleOwn[acc];
})) {
  throw new Error('页面VM对象中的属性data不可与\\"' + accessors.join(',') + '\\"同时存在，请使用private替换data名称');
} else if (!moduleOwn.data) {
  moduleOwn.data = {};
  moduleOwn._descriptor = {};
  accessors.forEach(function (acc) {
    const accType = typeof moduleOwn[acc];
    if (accType === 'object') {
      moduleOwn.data = Object.assign(moduleOwn.data, moduleOwn[acc]);
      for (const name in moduleOwn[acc]) {
        moduleOwn._descriptor[name] = {
          access: acc
        };
      }
    } else if (accType === 'function') {
      console.warn('页面VM对象中的属性' + acc + '的值不能是函数，请使用对象');
    }
  });
}}"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 7`] = `
"module.exports = {
  \\"type\\": \\"div\\",
  \\"attr\\": {},
  \\"classList\\": [
    \\"demo-page\\"
  ],
  \\"children\\": [
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": function () {return this.time}
      },
      \\"classList\\": [
        \\"title\\"
      ]
    },
    {
      \\"type\\": \\"input\\",
      \\"attr\\": {
        \\"type\\": \\"button\\",
        \\"value\\": \\"点击引入动态JS修改时间\\"
      },
      \\"classList\\": [
        \\"btn\\"
      ],
      \\"events\\": {
        \\"click\\": \\"dynamicImportJs\\"
      }
    }
  ]
}"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 8`] = `
"module.exports = {
  \\".demo-page\\": {
    \\"flexDirection\\": \\"column\\",
    \\"justifyContent\\": \\"center\\",
    \\"alignItems\\": \\"center\\"
  },
  \\".title\\": {
    \\"fontSize\\": \\"40px\\",
    \\"textAlign\\": \\"center\\"
  },
  \\".btn\\": {
    \\"width\\": \\"550px\\",
    \\"height\\": \\"86px\\",
    \\"marginTop\\": \\"75px\\",
    \\"borderRadius\\": \\"43px\\",
    \\"backgroundColor\\": \\"#09ba07\\",
    \\"fontSize\\": \\"30px\\",
    \\"color\\": \\"#ffffff\\"
  }
}"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 9`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {
  private: {}
};
exports.default = _default;
const moduleOwn = exports.default || module.exports;
const accessors = ['public', 'protected', 'private'];
if (moduleOwn.data && accessors.some(function (acc) {
  return moduleOwn[acc];
})) {
  throw new Error('页面VM对象中的属性data不可与\\"' + accessors.join(',') + '\\"同时存在，请使用private替换data名称');
} else if (!moduleOwn.data) {
  moduleOwn.data = {};
  moduleOwn._descriptor = {};
  accessors.forEach(function (acc) {
    const accType = typeof moduleOwn[acc];
    if (accType === 'object') {
      moduleOwn.data = Object.assign(moduleOwn.data, moduleOwn[acc]);
      for (const name in moduleOwn[acc]) {
        moduleOwn._descriptor[name] = {
          access: acc
        };
      }
    } else if (accType === 'function') {
      console.warn('页面VM对象中的属性' + acc + '的值不能是函数，请使用对象');
    }
  });
}}"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 10`] = `
"module.exports = {
  \\"type\\": \\"div\\",
  \\"attr\\": {}
}"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 11`] = `"module.exports = {}"`;

exports[`use dynamic import in project build results include split chunks for [SMART] 12`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.getGlobalData = getGlobalData;
exports.setGlobalData = setGlobalData;
/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * @file 全局能力的配置与获取
 */

function getGlobalRef() {
  return Object.getPrototypeOf(global) || global;
}
const quickappGlobal = getGlobalRef();

/**
 * 设置全局(被APP与Page共享)数据；
 * @param key {string}
 * @param val {*}
 */
function setGlobalData(key, val) {
  quickappGlobal[key] = val;
}

/**
 * 获取全局(被APP与Page共享)数据；
 * @param key {string}
 * @return {*}
 */
function getGlobalData(key) {
  return quickappGlobal[key];
}

// 两个方法默认定义在全局
setGlobalData('setGlobalData', setGlobalData);
setGlobalData('getGlobalData', getGlobalData);"
`;

exports[`use dynamic import in project build results include split chunks for [SMART] 13`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = exports.b = exports.a = void 0;
/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

const a = 1;
exports.a = a;
const b = 2;
exports.b = b;
function timerGenerator() {
  const time = new Date().toLocaleString();
  console.log(\`TIME: \${time}\`);
  return time;
}
var _default = timerGenerator;
exports.default = _default;"
`;
